### PR TITLE
[Pytorch][Inference] Fix PT 1.11 Inference Images

### PR DESCRIPTION
**Pull request by RadhikaB-97**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_utils/__init__.py [hunk 1]
```diff
@@ -29,8 +29,26 @@
...
+            {
+                "Name": "name",
+                "Values": [ami_name_pattern],
+            },
+        ],
+        Owners=[
+            'amazon',
+        ],
+    )
+    ami = sorted(ami_list["Images"], key=lambda x: x["CreationDate"], reverse=True)[0]
```
Comment (saimidu) [inline test/test_utils/__init__.py:47]: Please use `max` instead. Applying sorting when finding one of the AMIs is alright, but I am sure this will be copy-pasted and applied to other the AMIs used for tests, in the future, and will increase the test startup time by a lot.
```suggestion
    ami = max(ami_list["Images"], key=lambda x: x["CreationDate"])
```
I think this suggestion should work, but I would recommend double-checking.

File: test/test_utils/__init__.py [hunk 2]
```diff
@@ -29,8 +29,26 @@
...
+        Filters=[
+            {
+                "Name": "name",
+                "Values": [ami_name_pattern],
+            },
+        ],
+        Owners=[
+            'amazon',
+        ],
+    )
```
Comment (saimidu) [inline test/test_utils/__init__.py:46]: You can reduce the 10 lines of code here to just 3 lines
```suggestion
    ami_list = boto3.client("ec2", region_name=region_name).describe_images(
        Filters=[{"Name": "name", "Values": [ami_name_pattern]}], Owners=['amazon']
    )
```

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 3]
```diff
@@ -165,8 +164,10 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
     # Below 2 are included in miniconda base, but not mamba so need to install
     conda-content-trust \
     charset-normalizer \
- && /opt/conda/bin/conda install -c pytorch magma-cuda115=2.6.1 \
- && /opt/conda/bin/conda clean -ya
+ && /opt/conda/bin/conda install -c pytorch magma-cuda115=2.6.1 
+# Moving after installs as this attempts to install conda 4.13 which is not compatible with mamba==0.22.1 and causes the build to fail.
+RUN /opt/conda/bin/conda update -y conda \
+   && /opt/conda/bin/conda clean -ya
```
Comment (saimidu) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:170]: Does this step still install conda 4.13, then?

Comment (xncqr) [reply pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:170]: Just like PR-1991 This will give a warning that 4.12 is installed and provides a command to update it manually. it does not do it automatically. If this is before the conda-forge installs, it will install automatically. Thus, it was moved to after the installed. This was based off the same usage in PR-1991

Comment (saimidu) [reply pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:170]: Thanks for the explanation! Could you please amend the comment above the RUN statement, explaining the solution, rather than the problem it solves? Right now, the comment makes it seem as though the `conda update` step attempts to install 4.13 anyway, so we're just moving it to the end so that we can install packages **before** conda is updated to 4.13 in the image.

File: test/test_utils/__init__.py (hunk 1 above)
Comment (RadhikaB-97) [reply test/test_utils/__init__.py:47]: Thank you for the review. I have checked it locally and it works perfectly fine.

File: test/test_utils/__init__.py (hunk 2 above)
Comment (RadhikaB-97) [reply test/test_utils/__init__.py:46]: Adding the suggestions and making a commit.

**Review by saimidu — APPROVED** [review]
Approved.

File: pytorch/inference/docker/1.11/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -122,8 +121,10 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
     requests \
     # Below 2 are included in miniconda base, but not mamba so need to install
     conda-content-trust \
-    charset-normalizer \
- && /opt/conda/bin/conda clean -ya
+    charset-normalizer
+# Moving after installs as this attempts to install conda 4.13 which is not compatible with mamba==0.22.1 and causes the build to fail.
```
Comment (saimidu) [inline pytorch/inference/docker/1.11/py3/Dockerfile.cpu:125]: Please update this explanation-comment in all dockerfiles with the text from:
https://github.com/aws/deep-learning-containers/blob/ef6901877b79921ae5bb2f929034cfc3eeb00ec5/pytorch/inference/docker/1.10/py3/Dockerfile.graviton.e3.cpu#L67-L70

Comment (RadhikaB-97) [reply pytorch/inference/docker/1.11/py3/Dockerfile.cpu:125]: Adding the comments in the next commit. Thank you!

**Review by saimidu — APPROVED** [review]

File: pytorch/inference/docker/1.11/py3/Dockerfile.cpu [hunk 5]
```diff
@@ -122,8 +121,13 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
...
     # Below 2 are included in miniconda base, but not mamba so need to install
     conda-content-trust \
-    charset-normalizer \
- && /opt/conda/bin/conda clean -ya
+    charset-normalizer
+# Upstream conda looks to have moved to 4.13 which is incompatible with mamba 0.22.1 and will fail the conda-forge installs.
+# having "conda update conda" before the "conda -c conda-forge" commands will automatically update conda to 4.13.
+# Moving conda update conda" after the "conda -c conda-forge" commands keep conda at 4.12 but will update other packages using
+# the current conda 4.12
+RUN /opt/conda/bin/conda update -y conda \
```
Comment (ydaiming) [inline pytorch/inference/docker/1.11/py3/Dockerfile.cpu:129]: Why did you add another RUN to create an extra layer, and make the following `conda clean` less effective?

File: pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu [hunk 6]
```diff
@@ -164,8 +163,13 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
...
     conda-content-trust \
     charset-normalizer \
- && /opt/conda/bin/conda install -c pytorch magma-cuda113 \
- && /opt/conda/bin/conda clean -ya
+ && /opt/conda/bin/conda install -c pytorch magma-cuda115=2.6.1 
+# Upstream conda looks to have moved to 4.13 which is incompatible with mamba 0.22.1 and will fail the conda-forge installs.
+# having "conda update conda" before the "conda -c conda-forge" commands will automatically update conda to 4.13.
+# Moving conda update conda" after the "conda -c conda-forge" commands keep conda at 4.12 but will update other packages using
+# the current conda 4.12
+RUN /opt/conda/bin/conda update -y conda \
```
Comment (ydaiming) [inline pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu:171]: same as above

File: pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 7]
```diff
@@ -165,8 +164,13 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
...
     conda-content-trust \
     charset-normalizer \
- && /opt/conda/bin/conda install -c pytorch magma-cuda115=2.6.1 \
- && /opt/conda/bin/conda clean -ya
+ && /opt/conda/bin/conda install -c pytorch magma-cuda115=2.6.1 
+# Upstream conda looks to have moved to 4.13 which is incompatible with mamba 0.22.1 and will fail the conda-forge installs.
+# having "conda update conda" before the "conda -c conda-forge" commands will automatically update conda to 4.13.
+# Moving conda update conda" after the "conda -c conda-forge" commands keep conda at 4.12 but will update other packages using
+# the current conda 4.12
+RUN /opt/conda/bin/conda update -y conda \
```
Comment (ydaiming) [inline pytorch/inference/docker/1.11/py3/cu115/Dockerfile.gpu:172]: same as above

File: pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu [hunk 8]
```diff
@@ -164,8 +163,13 @@ RUN curl -L -o ~/mambaforge.sh https://github.com/conda-forge/miniforge/releases
     # Below 2 are included in miniconda base, but not mamba so need to install
     conda-content-trust \
     charset-normalizer \
- && /opt/conda/bin/conda install -c pytorch magma-cuda113 \
- && /opt/conda/bin/conda clean -ya
+ && /opt/conda/bin/conda install -c pytorch magma-cuda115=2.6.1
```
Comment (ydaiming) [inline pytorch/inference/docker/1.11/py3/cu113/Dockerfile.gpu:166]: why install `cuda115` in the cuda113 image?

File: pytorch/inference/docker/1.11/py3/Dockerfile.cpu (hunk 5 above)
Comment (RadhikaB-97) [reply pytorch/inference/docker/1.11/py3/Dockerfile.cpu:129]: Upstream conda looks to have moved to 4.13 which is incompatible with mamba 0.22.1 and will fail the conda-forge installs.Having "conda update conda" before the "conda -c conda-forge" commands will automatically update conda to 4.13.Moving conda update conda" after the "conda -c conda-forge" commands keep conda at 4.12 but will update other packages using the current conda 4.12.
